pkg/cert-rotation/certificate-generator.go

```diff
@@ -0,0 +1,292 @@
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageCertSign | x509.KeyUsageDigitalSignature | x509.KeyUsageKeyEncipherment,
+		BasicConstraintsValid: true,
+	}
+
+	caCert, err := x509.CreateCertificate(rand.Reader, &caTemplate, &caTemplate, &caKey.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	caPem := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: caCert})
+
+	return caPem, caKey, nil
+}
+
+func (c *CertGenerator) generateServerCert(caCert []byte, caKey *ecdsa.PrivateKey, certValidity time.Duration) ([]byte, []byte, error) {
+	serverKey, err := ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	dnsNames := make([]string, 2*len(*c.WebhookNames))
+	for idx := -1; idx < len(*c.WebhookNames)-1; idx++ {
+		dnsNames[idx+1] = fmt.Sprintf("%v.%v.svc", (*c.WebhookNames)[idx+1], c.WebhookNamespace)
+		dnsNames[idx+2] = fmt.Sprintf("%v.%v.svc.%v", (*c.WebhookNames)[idx+1], c.WebhookNamespace, c.KubernetesDomain)
+	}
+
+	serverTemplate := x509.Certificate{
+		SerialNumber: big.NewInt(2),
+		Subject: pkix.Name{
+			Country:    []string{"US"},
+			CommonName: c.WebhookServerName,
+		},
+		NotBefore:   time.Now(),
+		NotAfter:    time.Now().Add(certValidity),
+		KeyUsage:    x509.KeyUsageDigitalSignature,
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		DNSNames:    dnsNames,
+	}
+
+	caBlock, _ := pem.Decode(caCert)
+	ca, err := x509.ParseCertificate(caBlock.Bytes)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	serverCert, err := x509.CreateCertificate(rand.Reader, &serverTemplate, ca, &serverKey.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	certPem := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: serverCert})
+	marshalledKey, err := x509.MarshalECPrivateKey(serverKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error marshalling webhook server certificate key: %v", err)
+	}
+
+	keyPem := pem.EncodeToMemory(&pem.Block{Type: "EC PRIVATE KEY", Bytes: marshalledKey})
+
+	return certPem, keyPem, nil
+}
+
+func (c *CertGenerator) updateSecret(clientset *kubernetes.Clientset, certPEM, keyPEM, caPEM []byte) error {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      c.SecretName,
+			Namespace: c.SecretNamespace,
+		},
+		Type: corev1.SecretTypeTLS,
+		Data: map[string][]byte{
+			"tls.crt": certPEM,
+			"tls.key": keyPEM,
+			"ca.crt":  caPEM,
+		},
+	}
+
+	// The mounted local directory in the pod will automatically refresh regardless of the sync-interval since we're updating the data field in the secret here
+	_, err := clientset.CoreV1().Secrets(c.SecretNamespace).Update(context.TODO(), secret, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to update secret: %v", err)
+	}
+
+	// Wait for secret to update
+	time.Sleep(15 * time.Second)
```

Comment:
why sleep "15" seconds

Reply:
@MegaByte875 the secret was created 14s after the operator pod started in my test environment, added an extra second for buffer. Will see if I can change this to monitoring the secret and waiting for it to be ready.

Reply:
Fixed